source/layers/validation/checkers/events_deadlock/zel_events_deadlock_checker.cpp

```diff
@@ -0,0 +1,256 @@
+    uint32_t numWaitEvents,                  ///< [in][optional] number of events to wait on before launching; must be 0
+                                             ///< if `nullptr == phWaitEvents`
+    ze_event_handle_t *phWaitEvents          ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                             ///< on before launching
+) {
+
+    return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+eventsDeadlockChecker::ZEeventsDeadlockChecker::zeCommandListAppendLaunchCooperativeKernelEpilogue(
+    ze_command_list_handle_t hCommandList,   ///< [in] handle of the command list
+    ze_kernel_handle_t hKernel,              ///< [in] handle of the kernel object
+    const ze_group_count_t *pLaunchFuncArgs, ///< [in] thread group launch arguments
+    ze_event_handle_t hSignalEvent,          ///< [in][optional] handle of the event to signal on completion
+    uint32_t numWaitEvents,                  ///< [in][optional] number of events to wait on before launching; must be 0
+                                             ///< if `nullptr == phWaitEvents`
+    ze_event_handle_t *phWaitEvents          ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                             ///< on before launching
+) {
+
+    return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+eventsDeadlockChecker::ZEeventsDeadlockChecker::zeCommandListAppendMemoryCopyPrologue(
+    ze_command_list_handle_t hCommandList, ///< [in] handle of command list
+    void *dstptr,                          ///< [in] pointer to destination memory to copy to
+    const void *srcptr,                    ///< [in] pointer to source memory to copy from
+    size_t size,                           ///< [in] size in bytes to copy
+    ze_event_handle_t hSignalEvent,        ///< [in][optional] handle of the event to signal on completion
+    uint32_t numWaitEvents,                ///< [in][optional] number of events to wait on before launching; must be 0
+                                           ///< if `nullptr == phWaitEvents`
+    ze_event_handle_t *phWaitEvents        ///< [in][optional][range(0, numWaitEvents)] handle of the events to wait
+                                           ///< on before launching
+) {
+
+    std::cout << "eventsDeadlockChecker: zeCommandListAppendMemoryCopyPrologue" << std::endl;
+    printf("\t\tNumber of wait events: %d\n", numWaitEvents);
+    for (uint32_t i = 0; i < numWaitEvents; i++) {
+        printf("\t\tphWaitEvents[%d] pointer address: %p\n", i, phWaitEvents[i]);
+    }
+
+    int this_action_new_node_id = invalidDagID;
+
+    if (hSignalEvent != nullptr) {
+
+        if (eventToDagID.find(hSignalEvent) != eventToDagID.end()) {
+            if (eventToDagID[hSignalEvent] != invalidDagID) {
+                // This event already exists in the DAG. Get the DAG node ID.
```

Comment:
just a nitpick, it's not really a DAG if there is a deadlock - because then there is a cycle.

Reply:
Yes, when a cycle is added the underlying DAG structure will not accept that `.add()` and will return error. This way we know there is a cycle and hence a deadlock.